internal/rules/resource_management.go

```diff
@@ -49,11 +73,15 @@ type ResourceManagement interface {
 	GarbageCollectServiceAccounts(
 		ctx context.Context, expectedServiceAccounts *set.Set[string], targetNamespaces *set.Set[string],
 	) (ctrl.Result, error)
+	GarbageCollectRoles(ctx context.Context, resources []WSAResource) error
```

Comment:
I think we've seen a couple of times that the role/clusterrole almost duplication can be confusing - explicitness might be our saviour here

```suggestion
	GarbageCollectRolesAndClusterRoles(ctx context.Context, resources []WSAResource) error
```

Reply:
Or split these two methods up